### ISSUE TITLE
[LaTeX] figure floated to next page sometimes leaves extra vertical whitespace

### ISSUE THREAD

**Issue by jessetan** [sphinx-doc/sphinx#3881]
This is a cornercase when generating PDF through LaTeX. I managed to track it down as specifically as described here, but I don't know if this is an issue of Sphinx or TeX Live.

### Problem
When a figure does not fit on the page, it is floated to the top of the next page. The place where the original `figure` directive was will effectively become a paragraph break and have the regular vertical spacing (`parskip`?)  _except_ if the width of the last line of the preceding paragraph is the same as the line width. In that case it looks to be double.

#### Procedure to reproduce the problem
Use the attached rst file to create a pdf. It uses `sphinxheader.png` from the Sphinx homepage but any other suffciently high enough image should have the same effect.

#### Error logs / results

Result I see: 
![extra space](https://user-images.githubusercontent.com/3284111/27277948-487a845e-54e0-11e7-9062-af16b827e1b6.png)

Making the last line one character shorter gives the expected result:
![one char less](https://user-images.githubusercontent.com/3284111/27277967-5d296fdc-54e0-11e7-9c19-cf05d56b0773.png)

As does adding one character more:
![one char more](https://user-images.githubusercontent.com/3284111/27277975-665a9928-54e0-11e7-915d-74d1e89ff70d.png)

### Reproducible project / your project
[figurespace.zip](https://github.com/sphinx-doc/sphinx/files/1084457/figurespace.zip)

### Environment info
- OS: MacOS 10.12.5
- Python version: 2.7.13
- Sphinx version: 1.5.5
- TeX Live version: 2016

**Comment by jfbu** [comment]
Good catch, this is both a LaTeX, even a TeX issue (which is subtle) and possibly a Sphinx issue.

First to reproduce your problem I use this `index.rst` 

```
TITLE
=====

.. raw:: latex

   \null\clearpage

Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

.. raw:: latex

   \vspace*{\dimexpr\textheight-2.5cm}

AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA  IIIIAAIII.

.. figure:: sphinxheader.png

Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

and 

```
latex_elements = {
    # The paper size ('letterpaper' or 'a4paper').
    #
    'papersize': 'letterpaper',
}

latex_documents = [
    (master_doc, 'figurespace.tex', u'figurespace Documentation',
     u'dsf', 'howto'),
]
```

in conf.py (so that I can get my PDF viewer to display the page 2 and 3 side to side and confirm the effect).

I had encountered a bit analogous problem with this kind of latex file:

```
\documentclass{article}

\begin{document}

% example edited because the \noindent were missing, making it less demonstrative of the issue
\noindent\begin{minipage}{\linewidth}
  foo
\end{minipage}
{}

bar

\noindent\begin{minipage}{\linewidth}
  foo
\end{minipage}

bar


\end{document}
```

If you compile this with pdflatex you will see a strange extra white line above the first bar.

Other types of situation cause this (especially when the minipage is surrounded in an environment not doing ignorespacesafterend).

Now your example looked a bit different because it involves a page break (and I fine tuned my minimal example to use it.) However, the page break is irrelevant: take out the 

```
.. raw:: latex

   \vspace*{\dimexpr\textheight-2.5cm}

```

and the same phenomenon is observed. Thus your situation is


```
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA  IIIIAAIII.
\begin{figure}[htbp]
\centering

\noindent\sphinxincludegraphics{{sphinxheader}.png}
\end{figure}
```

in letterpaper size, and the AAAA... line fills exactly the linewidth with limited stretch/shrink (only one space).

I had not seen this before, but it belongs to the category above. I know that this is not a pure LaTeX issue, it can be created with only TeX primitives.

This problem disappears (and the two options give identical output) either with

```
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA  IIIIAAIII.%
\begin{figure}[htbp]
\centering

\noindent\sphinxincludegraphics{{sphinxheader}.png}
\end{figure}
```

or with

```
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA  IIIIAAIII.

\begin{figure}[htbp]
\centering

\noindent\sphinxincludegraphics{{sphinxheader}.png}
\end{figure}
```


Arguably, Sphinx should do the latter because a `figure` is block-level element in LaTeX. However it is possible to use `H` specifier, and perhaps some examination is needed.

For the record, with the "bad" input, one gets this extra at top of page in TeX trace:

```
...\vbox(650.43001+0.0)x469.75502, glue set 487.09692fil
....\glue(\topskip) 10.0
....\hbox(0.0+0.0)x469.75502, glue set 469.75502fil
.....\penalty 10000
.....\penalty 10000
.....\glue(\parfillskip) 0.0 plus 1.0fil
.....\glue(\rightskip) 0.0
....\penalty 0
....\glue 12.0 plus 2.0 minus 2.0
....\vbox(75.28105+0.0)x469.75502
```

which shows that there is 10pt + 12 pt extra blank space on top of new page, compared to "good" situation

```
...\vbox(650.43001+0.0)x469.75502, glue set 509.09457fil
....\vbox(75.28105+0.0)x469.75502
```

(the `\vbox(75.28105+0.0)x469.75502` is the included image, and it is injected into the page via TeX `\insert` mechanism. It's all voodoo.)

In the bad case TeX inserts a `\hbox{}` and only then the figure, thus using its `h`(ere)  parameter, not an insert at top of page. One can confirm this analysis again with

```
\vspace*{\dimexpr\textheight-2.5cm}

AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA  IIIIAAIII.
\begin{figure}[t]
\centering

\noindent\sphinxincludegraphics{{sphinxheader}.png}
\end{figure}
``` 

it forces the insertion at top of page, hence the extraneous space disappears (but it is **again there** _after_ the figure).

Anyway I just tested now with H specifier: in that case the problem does not exist in the first place, and adding a blank line above figure does not modify output.

Thus we will fix this by adding a blank line before `figure` always. Which makes sense because `figure` in LaTeX is definitely block-level, hence should be properly separated from previous paragraph by blank line in latex source for readability.

At the root of the matter there seems to be some bug, if I dare say, of the TeX paragraph building routine, regarding space tokens. In some cases an extra space token counter-intuitively leads TeX to end a paragraph with an empty `\hbox` of full width and empty contents.

Thanks for reporting!

**Comment by jfbu** [comment]
Fixed at a069378a.

addendum: in LaTeX it is licit to put `\begin{figure}... \end{figure}`  as an "inline" element. It will then try to appear right there in middle of paragraph, lines being still fully justified. However reStructured text will treat figure as block-level. The LaTeX writer does `\end{figure}\n` and text after it will insert a `\n`  at paragraph start. So we end-up always with blank line in tex source after `\end{figure}`.  In those cases, this issue reported by @jessetan demonstrates LaTeX `figure` has a problem, which is correctly corrected (sic) by also having  a blank line before `\begin{figure}` in tex source. At least that's my conclusion. I must admit I had never used `figure` in latex without blank lines in source before and after. I didn't know LaTeX made provisions for use in middle of paragraph. But this is not possible for Sphinx anyhow.
